db/bbolt.go

```diff
@@ -0,0 +1,262 @@
+		bb.logger.Error("Error getting latest block", zap.Error(err))
+		return err
+	}
+
+	// Update latest block if it's the latest
+	err = bb.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(latestBlockBucket))
+		if err != nil {
+			bb.logger.Error("Error getting latest block", zap.Error(err))
+			return err
+		}
+		if latestBlock.BlockHeight < block.BlockHeight {
+			return b.Put([]byte(latestBlockKey), bb.itob(block.BlockHeight))
+		}
+		return nil
+	})
+	if err != nil {
+		bb.logger.Error("Error updating latest block", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func (bb *BBoltHandler) GetBlockByHeight(height uint64) (*types.Block, error) {
+	var block types.Block
+	err := bb.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(blocksBucket))
+		v := b.Get(bb.itob(height))
+		if v == nil {
+			return types.ErrBlockNotFound
+		}
+		return json.Unmarshal(v, &block)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &block, nil
+}
+
+func (bb *BBoltHandler) GetBlockByHash(hash string) (*types.Block, error) {
+	// Fetch block number corresponding to hash
+	var blockHeight uint64
+	err := bb.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(blockHeightsBucket))
+		v := b.Get([]byte(hash))
+		if v == nil {
+			return types.ErrBlockNotFound
+		}
+		blockHeight = bb.btoi(v)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return bb.GetBlockByHeight(blockHeight)
+}
+
+func (bb *BBoltHandler) GetBlockStatusByHeight(height uint64) (bool, error) {
+	_, err := bb.GetBlockByHeight(height)
+	if err != nil {
+		if errors.Is(err, types.ErrBlockNotFound) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (bb *BBoltHandler) GetBlockStatusByHash(hash string) (bool, error) {
+	// Fetch block number by hash
+	var blockHeight uint64
+	err := bb.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(blockHeightsBucket))
+		res := b.Get([]byte(hash))
+		if len(res) == 0 {
+			return types.ErrBlockNotFound
+		}
+		blockHeight = bb.btoi(res)
+		return nil
+	})
+	if err != nil {
+		if errors.Is(err, types.ErrBlockNotFound) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return bb.GetBlockStatusByHeight(blockHeight)
+}
+
+func (bb *BBoltHandler) GetLatestBlock() (*types.Block, error) {
+	var latestBlockHeight uint64
+
+	// Fetch latest block height
+	err := bb.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(latestBlockBucket))
+		v := b.Get([]byte(latestBlockKey))
+		if v == nil {
+			return types.ErrBlockNotFound
+		}
+		latestBlockHeight = bb.btoi(v)
+		return nil
+	})
+	if err != nil {
+		// If no latest block has been stored yet, return empty block (block 0)
+		if errors.Is(err, types.ErrBlockNotFound) {
+			return nil, nil
+		}
+		bb.logger.Error("Error getting latest block", zap.Error(err))
+		return nil, err
+	}
+
+	// Fetch latest block by height
+	return bb.GetBlockByHeight(latestBlockHeight)
+}
+
+func (bb *BBoltHandler) DeleteDB() error {
```

Comment:
i think this method is too dangerous. i'd recommend we get rid of this API. so by default the DB is consistent

if people want to delete DB, they can remove call `rm` themselves